### ISSUE TITLE
Problem doing file.managed with a http source

### ISSUE THREAD

**Issue by davidkarlsen** [saltstack/salt#28641]
OS: https://vagrantcloud.com/boxcutter/boxes/oel67/versions/3.0.3

State:
```
/tmp/p17694377_121020_Linux-x86-64_1of8.zip:
  file.managed:
    - source: http://myserver.mydomain/vagrant_boxes/sw/oracle_source/p17694377_121020_Linux-x86-64_1of8.zip
    - source_hash: md5=017d624abe5165406d3d7c7d2d3a834c
```

```
salt version:
15:18:34 Checking if salt-minion is installed
15:18:34 salt-minion was not found.
15:18:34 Checking if salt-call is installed
15:18:34 salt-call was not found.
15:18:34 Using Bootstrap Options: 
15:18:34 Bootstrapping Salt... (this may take a while)
15:18:35 
15:18:35 
15:18:35 %
15:18:35 
15:18:35 T
15:18:35 otal    % Received % Xferd  Average Speed   Time    Time     Time  Current
15:18:35                                  Dload  Upload   Total   Spent    Left  Speed
15:18:35 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
15:18:36 
15:18:36 
15:18:36 
15:18:36 0
15:18:36      0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
15:18:36  *  INFO:  sh -- Version 2015.11.04
15:18:36  *  INFO: System Information:
15:18:36  *  INFO:   CPU:          GenuineIntel
15:18:36  *  INFO:   CPU Arch:     x86_64
15:18:36  *  INFO:   OS Name:      Linux
15:18:36  *  INFO:   OS Version:   3.8.13-68.3.4.el6uek.x86_64
15:18:36  *  INFO:   Distribution: Oracle Linux 6.7
15:18:36 
15:18:36  *  INFO: Installing minion
15:18:36 
100  209k  100  209k    0     0   140k      0  0:00:01  0:00:01 --:--:--  159k
15:18:36  *  INFO: Found function install_oracle_linux_stable_deps
15:18:36  *  INFO: Found function install_oracle_linux_stable
15:18:36  *  INFO: Found function install_oracle_linux_stable_post
15:18:36  *  INFO: Found function install_oracle_linux_restart_daemons
15:18:36  *  INFO: Found function daemons_running
15:18:36  *  INFO: Found function install_oracle_linux_check_services
15:18:36  *  INFO: Running install_oracle_linux_stable_deps()
15:18:37  *  INFO: Testing if packages usually on the optionals repository are available:
15:18:37  *  INFO:   - python-jinja2
```

```
INFO    ] Running state [/tmp/p17694377_121020_Linux-x86-64_1of8.zip] at time 14:19:51.421276
15:19:52 ==> default: [INFO    ] Executing state file.managed for /tmp/p17694377_121020_Linux-x86-64_1of8.zip
15:21:35 ==> default: [ERROR   ] Uncaught exception
15:21:35 ==> default: Traceback (most recent call last):
15:21:35 ==> default:   File "/usr/lib64/python2.6/site-packages/tornado/http1connection.py", line 237, in _read_message
15:21:35 ==> default:     delegate.finish()
15:21:35 ==> default:   File "/usr/lib64/python2.6/site-packages/tornado/http1connection.py", line 637, in finish
15:21:35 ==> default:     return self._delegate.finish()
15:21:35 ==> default:   File "/usr/lib64/python2.6/site-packages/tornado/simple_httpclient.py", line 463, in finish
15:21:35 ==> default:     data = b''.join(self.chunks)
15:21:35 ==> default: MemoryError
15:21:35 ==> default: [ERROR   ] Unable to manage file: 
15:21:35 ==> default: [INFO    ] Completed state [/tmp/p17694377_121020_Linux-x86-64_1of8.zip] at time 14:21:34.238356
15:21:35 ==> default: [INFO    ] Running state [/tmp/p17694377_121020_Linux-x86-64_2of8.zip] at time 14:21:34.257553
15:21:35 ==> default: [INFO    ] Executing state file.managed for /tmp/p17694377_121020_Linux-x86-64_2of8.zip
15:22:39 ==> default: [ERROR   ] Uncaught exception
15:22:39 ==> default: Traceback (most recent call last):
15:22:39 ==> default:   File "/usr/lib64/python2.6/site-packages/tornado/http1connection.py", line 237, in _read_message
15:22:39 ==> default:     delegate.finish()
15:22:39 ==> default:   File "/usr/lib64/python2.6/site-packages/tornado/http1connection.py", line 637, in finish
15:22:39 ==> default:     return self._delegate.finish()
15:22:39 ==> default:   File "/usr/lib64/python2.6/site-packages/tornado/simple_httpclient.py", line 502, in finish
15:22:39 ==> default:     buffer = BytesIO(data)  # TODO: don't require one big string?
15:22:39 ==> default: MemoryError
```

**Comment by jfindlay** [comment]
@davidkarlsen, thanks for the report.  This is something that has been fixed in later releases.  Your bootstrap output doesn't show what salt version you are running, but upgrading to 2015.8.1 should fix this.

**Comment by davidkarlsen** [comment]
@jfindlay See above, it says "15:18:36  *  INFO:  sh -- Version 2015.11.04"

**Comment by davidkarlsen** [comment]
@jfindlay ping?

**Comment by anlutro** [comment]
2015.11 isn't even a real version, is it?

**Comment by jfindlay** [comment]
@davidkarlsen, the version reported in the bootstrap output is the version of bootstrap itself.  Can you paste the output of `salt --versions-report` or similar?  Thanks.

**Comment by bl4ckcontact** [comment]
I'm getting the same thing copying a file with file.managed via HTTP.

```
Salt Version:
           Salt: 2015.8.1

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: Not Installed
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.7.0
         Python: 2.7.10 (default, May 23 2015, 09:44:00) [MSC v.1500 64 bit (AMD64)]
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.2
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: 0.3.7
          smmap: Not Installed
        timelib: Not Installed

System Versions:
           dist:
        machine: AMD64
        release: 2008ServerR2
         system: 2008ServerR2 6.1.7601 SP1 Multiprocessor Free
```

**Comment by cachedout** [comment]
This is definitely not fixed. I'm not sure where that came from.

The problem here is that that Tornado seems to be spooling the entire file into memory. The likely fix here is to add a streaming callback into Salt's file client so that we can write out the file in chunks.

**Comment by benhosmer** [comment]
I get something similar, but a slightly different tornado error:

```
Salt Version:
           Salt: 2015.8.3
 
Dependency Versions:
         Jinja2: unknown
       M2Crypto: 0.20.2
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
         Python: 2.6.6 (r266:84292, Jul 23 2015, 15:22:56)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed
 
System Versions:
           dist: centos 6.6 Final
        machine: x86_64
        release: 2.6.32-504.3.3.el6.x86_64
         system: CentOS 6.6 Final
```

```
[INFO    ] Executing state file.managed for /data/omar/bmng/world.200406.A1.tif
[ERROR   ] Uncaught exception
Traceback (most recent call last):
  File "/usr/lib64/python2.6/site-packages/tornado/http1connection.py", line 237, in _read_message
    delegate.finish()
  File "/usr/lib64/python2.6/site-packages/tornado/http1connection.py", line 637, in finish
    return self._delegate.finish()
  File "/usr/lib64/python2.6/site-packages/tornado/simple_httpclient.py", line 502, in finish
    buffer = BytesIO(data)  # TODO: don't require one big string?
MemoryError
[ERROR   ] Unable to manage file: 
[INFO    ] Completed state [/data/omar/bmng/world.200406.A1.tif] at time 13:27:29.872248
```

**Comment by benhosmer** [comment]
@cachedout It's been a while since I've been in the codebase, can you point me to where this should be implemented?

**Comment by cachedout** [comment]
@benhosmer Funny you should ask, since I think @DmitryKuzmenko may have fixed this earlier today. See #30704 for reference.

**Comment by benhosmer** [comment]
Beautiful! Thanks for following up so quickly @cachedout

**Comment by DmitryKuzmenko** [comment]
There are two issues with fetching files:

1. Non-chunked fetch with Tornado simple_httpclient running out of memory for large files (> RAM avail) because it fetches the entire content into a memory buffer.

2. Chunked fetch has upstream tornado related issue: redirects bodies are placed into the file together with the real file content. So fetched file became broken.

Here's the info about what Salt versions contain what issues:
2015.8.0 - 1: fixed, 2: present
2015.8.1..4 - 1: present, 2: fixed
develop - 1: fixed, 2: fixed

**Comment by benhosmer** [comment]
Thanks @DmitryKuzmenko for updating this so quickly.

**Comment by blbradley** [comment]
Is this in 2015.8.7?

**Comment by DmitryKuzmenko** [comment]
@blbradley unfortunately it's in the only 2016.3 at this moment. Do you have any cons about the upgrade ot the latest version?

**Comment by blbradley** [comment]
@DmitryKuzmenko I added more memory to the affected node. Thanks for your concern.

**Comment by DmitryKuzmenko** [comment]
The fix for this issue is #30704.
Just adding to be linked.
